src/bokeh/plotting/_tools.py

```diff
@@ -191,21 +191,26 @@ def _resolve_tools(tools: str | Sequence[Tool | str]) -> tuple[list[Tool], dict[
 
     return tool_objs, tool_map
 
-def _collect_repeated_tools(tool_objs: list[Tool]) -> Iterator[Tool]:
-    @dataclass(frozen=True)
-    class Item:
-        obj: Tool
-        properties: dict[str, Any]
+@dataclass(frozen=True)
+class Item:
+    obj: Tool
+    properties: dict[str, Any]
 
+def _collect_repeated_tools(tool_objs: list[Tool]) -> Iterator[Tool]:
     key: Callable[[Tool], str] = lambda obj: obj.__class__.__name__
-
     for _, group in itertools.groupby(sorted(tool_objs, key=key), key=key):
-        rest = [ Item(obj, obj.properties_with_values()) for obj in group ]
-        while len(rest) > 1:
-            head, *rest = rest
-            for item in rest:
-                if item.properties == head.properties:
-                    yield item.obj
+        grouped = list(group)
+        n = len(grouped)
+        if n > 1:
+            props = [Item(obj, obj.properties_with_values()) for obj in grouped]
+            i = 0
+            while i < len(props) - 1:
+                head = props[i]
+                for j in range(i+1, len(props)):
+                    item = props[j]
+                    if item.properties == head.properties:
+                        yield item.obj
+                i += 1
```

Comment:
```suggestion
            for i in range(len(props)):
                head = props[i]
                for j in range(i+1, len(props)):
                    item = props[j]
                    if item.properties == head.properties:
                        yield item.obj
```

IMHO this is more readable and shouldn't be slower.